packages/dds/tree/src/tree/anchorSet.ts

```diff
@@ -212,6 +212,8 @@ export class AnchorSet {
                 // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                 srcChildren[index].parentIndex < srcStart!.parentIndex + count
             ) {
+                // Sever the child -/> parent connection
+                srcChildren[index].parentPath = undefined;
```

Comment:
I had indented this code to avoid updating the up pointers twice, and just rely on the setting of the parent pointers below where they are attached to the destination.

Nodes which are deleted (and thus not attached to a destination) were can instead be handled in the dst == undefined case (and can be marked as deleted by deepDelete). At that point their parent info should be irrelevant and does not need to be cleared, but if you do want to clear it, doing it recursivly in deepDelete would make more sense.

Thus I'm unclear why this change is needed.

Reply:
The problem with not setting the `parentPath` to `undefined` in the delete case is that we then end up with nodes that try to call `this.parentPath?.removeChild(this)` when they are disposed despite their parent having already purged them.

I considered having the implementation of `removeChild` allow calls where the child is not present in the `children` list when the child's status is "Deleted", but that seems like it weakens our ability to detect bugs, and wastes time performing doomed searches.

Another option would be to not call `this.parentPath?.removeChild(this)` during dispose if the node is deleted, but not all deleted nodes have the parent -> child connection severed (only those nodes that are if the field where the deletion happened). So we'd end up leaking those descendants until the deleted nodes were disposed.

Yet another option would be to have a different status flag for nodes that were directly deleted from nodes whose ancestor were deleted. But...

...overall, it seemed simpler to have the invariant that the parent -> child and child -> parent connections are kept in sync.
The code seems much easier to reason about this way.

Reply:
As noted below, if we simply don't call this.parentPath?.removeChild(this) on nodes that are already dead (and thus illegal to access according to docs on Status.Dead), we do less work on anchor freeing, and less work here, and fix the bug, right?

Reply:
I dont think there is a real leak in that case: we have GC that can collect cycles, and if that was a problem we could just delete all the child and parent referenced from the nodes when marking them as deleted: that would be faster than trying to update them for each child one at a time

Reply:
Invariant wise, I find the invariant that for nodes that are dead or freed, thir data is unspecified and never touched, except for the status flag to be simplest. Thats what the docs on status are trying to say, and allows us to optimize by not caring about any of the other fields in these cases.

Reply:
I agree it's not a real leak (that's what I meant with the "until the deleted nodes were disposed" part).

> if we simply don't call this.parentPath?.removeChild(this) on nodes that are already dead (and thus illegal to access according to docs on Status.Dead), we do less work on anchor freeing, and less work here, and fix the bug, right?

I think that's correct, but there's two drawbacks:
1. We lose the ability to detect some usage bugs in the client code.
2. We end up with a more complex invariant when it comes to the reciprocity of `parentPath` and `children`.

This code is going to undergo some changes when we make it field-kinds specific and add more capabilities for its use in change rebasing. I can't make the case that those changes would suffer from the approach you propose, but I'd rather retain the simpler model. We can always make it more complex later.

Reply:
I don't see either of those drawpacks.

1. What client usage can we fail to reject? The internal anchors we might not free are only ones that were never exposed in the client API: if they were exposed, they would have and entry in the anchor map and bugs related to not freeing them can be detected there and well as actually freed.

2. I think the invariant that all Alive Nodes form a doubly linked tree rotted at the root, and deleted /freed nodes are not part of that tree and cannot be used other than their status for the purpose of assenting on double frees is a clean invariant compared to the other options.

Reply:
I forgot about the anchor map.
You've convinced me that the resulting invariant is not so bad.
I've changed the name of `Status.Dead` to `Status.Dangling`, which I think is less confusing.